packages/block-library/src/image/edit.js

```diff
@@ -345,6 +349,80 @@ export function ImageEdit( {
 		className: classes,
 	} );
 
+	async function onHTMLDrop( HTML ) {
+		const blocks = pasteHandler( { HTML, mode: 'BLOCKS' } );
+
+		if ( ! blocks || blocks.length !== 1 ) {
+			return;
+		}
+
+		const [ block ] = blocks;
+
+		if ( ! block || block.name !== 'core/image' ) {
+			return;
+		}
+
+		// Optimistic update to replace the current image.
+		setAttributes( {
+			url: undefined,
+			alt: undefined,
+			id: undefined,
+			title: undefined,
+			caption: undefined,
+			...block.attributes,
+		} );
+
+		// Media item already exists in the library.
+		if ( !! block.attributes.id ) {
+			selectBlock( clientId );
+			return;
+		}
+
+		// Media item doesn't exist in the library, upload it so that it
```

Comment:
The reason I didn't add such specific logic in my PR is:
1. We'll probably need to augment the media inserter API with the block to insert(could be third party blocks). That's why it'd be ideal for media tab and the drop zones to have a way to communicate or a different approach like setting the `dataTransfer` data to multiple types so transforms could handle them easier.
2. This should be the case soon for other types of media(video,audio), so similar code would need to be added to all of these blocks.